File: perl/search/get_granules.pl
#!/usr/bin/env perl

use strict;
use warnings;

use HTTP::Status qw(:constants);
use REST::Client;
use URI::Encode qw(uri_encode);
use XML::Simple;

sub get_token {
  my ($username, $password) = @_;

  # The REST endpoint to create a token in the ECHO Testbed environment
  my $echo_rest_tokens_url = 'https://testbed.echo.nasa.gov/echo-rest/tokens';

  # Create the REST client
  my $client = REST::Client->new();

  # We'll be using XML here.  We could also have used application/json
  my $request_headers = {
    'Content-Type' => 'application/xml',
    'Accept' => 'application/xml'
  };

  # The token request being sent
  my $token_request_hash_ref = {
    'username' => { 'content' => $username },
    'password' => { 'content' => $password  },
    'client_id' => { 'content' => 'ETIM demo' },
    'user_ip_address' => { 'content' => '127.0.0.1' }
  };

  my $token_request_xml = XMLout($token_request_hash_ref, RootName => 'token');

  # POST the request to the REST API
  $client->POST($echo_rest_tokens_url, $token_request_xml, $request_headers);

  my $token = undef;

  if ($client->responseCode() == HTTP_CREATED) {
    # If we get a 201 status code back, the token was created
    my $response_hash_ref = XMLin($client->responseContent());
    $token = $response_hash_ref->{'id'};
  }
  else {
    # If we get anything other than a 201, something went wrong
    print "Failed to create token\n";
    parse_and_display_errors($client->responseContent());
    exit 1;
  }

  return $token;
}

sub parse_and_display_errors {
  my $error_xml = shift;

  my $errors_ref = XMLin($error_xml, ForceArray => qw(error));

  foreach my $error (@{$errors_ref->{'error'}}) { print "  - ${error}\n"; }

  return;
}

sub get_dataset_id {
  my $token = shift;
  my $datasets_url = 'https://testbed.echo.nasa.gov/catalog-rest/echo_catalog/datasets';

  # Build up the list of search parameters that we're using
  my @query_parameters = ();
  push(@query_parameters, 'bounding_box=10.488,-0.703,53.331,68.906');
  push(@query_parameters, 'temporal[]=2009-01-01T10:00:00Z,2010-03-10T12:00:00Z');
  push(@query_parameters, 'provider=LPDAAC_ECS');

  my $query_string = uri_encode(join(q{&}, @query_parameters));

  my $request_url = "${datasets_url}?${query_string}";

  # Create the REST client
  my $client = REST::Client->new();

  # We'll be using XML here.  We could also have used application/json
  my $request_headers = {
    'Accept' => 'application/xml',
    'Echo-Token'=> $token
  };

  # Issue the GET request
  $client->GET($request_url, $request_headers);

  my $response_hash_ref = XMLin($client->responseContent());
  return $response_hash_ref->{'reference'}->{'id'};
}

sub get_granules {
  my ($token, $dataset_id) = @_;
  my $granules_url = 'https://testbed.echo.nasa.gov/catalog-rest/echo_catalog/granules';

  # Build up the list of search parameters that we're using
  my @query_parameters = ();
  push(@query_parameters, 'page_size=10');
  push(@query_parameters, 'page_num=1');
  push(@query_parameters, 'bounding_box=10.488,-0.703,53.331,68.906');
  push(@query_parameters, 'temporal[]=2009-01-01T10:00:00Z,2010-03-10T12:00:00Z');
  push(@query_parameters, 'provider=LPDAAC_ECS');
  push(@query_parameters, "echo_collection_id=${dataset_id}");

  my $query_string = join(q{&}, @query_parameters);

  my $request_url = "${granules_url}?${query_string}";

  # Create the REST client
  my $client = REST::Client->new();

  # We'll be using XML here.  We could also have used application/json
  my $request_headers = {
    'Accept' => 'application/echo10+xml',
    'Echo-Token'=> $token
  };

  # Issue the GET request
  $client->GET($request_url, $request_headers);

  return XMLin($client->responseContent());
}

my $token = get_token('guest', 'woot');
my $dataset_id = get_dataset_id($token);
my $granules = get_granules($token, $dataset_id);

print "Found these granule IDs:\n";
foreach my $granule (@{$granules->{'result'}}) {
  print $granule->{'echo_granule_id'} . "\n";
}

File: java/search/GetGranules.java
import java.io.*;
import java.net.URI;

import org.apache.http.client.utils.URIBuilder;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.DefaultHttpClient;


/**
 * This example shows how to search for and retrieve granule metadata from ECHO using 
 * the Catalog REST API.
 * 
 * @author yliu
 */
public class GetGranules
{
  public static void main(String[] args) throws Exception
  {
  	String format = null;
  	if (args.length > 0)
  	{
  		format = args[0];
  	}

  	// Create the HTTPClient
    DefaultHttpClient httpClient = new DefaultHttpClient();

    String reference = getGranules(httpClient, format);
    System.out.println(reference);
  }

  /**
   * Retrieve the granules in the given format.
   * 
   * @param httpClient
   *          the http client to use to send the data
   * @param format
   *          the format to retrieve the granules with
   * @return the granules in the given format.
   * @throws Exception
   */
  private static String getGranules(HttpClient httpClient, String format) throws Exception
  {
  	String url = "http://testbed.echo.nasa.gov/catalog-rest/echo_catalog/granules";
  	if (format != null)
  	{
  		url += "." + format;
  	}
  	
    URI uri = new URI(url);
    URIBuilder uribuilder = new URIBuilder(uri);
    uribuilder.setParameter("dataset_id", "MODIS/Terra+Aqua Nadir BRDF-Adjusted Reflectance 16-Day L3 Global 500m SIN Grid V005");
    uribuilder.setParameter("bounding_box", "10.488,-0.703,53.331,68.906");
    uribuilder.setParameter("temporal[]", "2009-01-01T10:00:00Z,2010-03-10T12:00:00Z");
    uribuilder.setParameter("provider", "LPDAAC_ECS");
    URI request_url = uribuilder.build();
    System.out.println("Request URL: " + request_url.toString());

    HttpGet httpGet = new HttpGet(request_url);
    HttpResponse response = httpClient.execute(httpGet);

    int responseCode = response.getStatusLine().getStatusCode();
    if (responseCode != 200)
    {
      System.out.println("Unexpected response code: " + responseCode);
    }

    String reference;
    try
    {
      reference = readInputStream(response.getEntity().getContent());
    }
    finally
    {
      httpGet.releaseConnection();
    }
    return reference;
  }

  /**
   * Reads an input stream and returns the contents as a string.
   * 
   * @param is
   *          input stream
   * @return string contents of input string
   * @throws IOException
   */
  private static String readInputStream(InputStream is) throws IOException
  {
    LineNumberReader reader = new LineNumberReader(new InputStreamReader(is));
    try
    {
      StringBuilder builder = new StringBuilder();
      String line = null;
      while ((line = reader.readLine()) != null)
      {
        if (builder.length() != 0)
        {
          builder.append("\n");
        }
        builder.append(line);
      }
      return builder.toString();
    }
    finally
    {
      reader.close();
    }
  }

}